### PR TITLE
Add common set theory filters.

### PR DESCRIPTION
**Pull request by mattcoley**
Fixes https://github.com/HubSpot/jinjava/issues/261
Adds `union`, `intersect`, `difference`, and `symmetric_difference` similar to Ansible https://ansible-docs.readthedocs.io/zh/stable-2.0/rst/playbooks_filters.html#set-theory-filters.

### CODE REVIEW

File: src/main/java/com/hubspot/jinjava/lib/filter/DifferenceFilter.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+import com.hubspot.jinjava.doc.annotations.JinjavaDoc;
+import com.hubspot.jinjava.doc.annotations.JinjavaParam;
+import com.hubspot.jinjava.doc.annotations.JinjavaSnippet;
+import com.hubspot.jinjava.interpret.JinjavaInterpreter;
+import com.hubspot.jinjava.util.ForLoop;
+import com.hubspot.jinjava.util.ObjectIterator;
+
+@JinjavaDoc(
+    value = "Returns a list containing elements present in the first list and not the second list",
```
Comment (boulter) [inline src/main/java/com/hubspot/jinjava/lib/filter/DifferenceFilter.java]: _but_ not

File: src/main/java/com/hubspot/jinjava/lib/filter/UnionFilter.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    snippets = {
+        @JinjavaSnippet(
+            code = "{{ [1, 2, 3]|union([2, 3, 4]) }}")
+    })
+public class UnionFilter implements AdvancedFilter {
+
+  @Override
+  public Object filter(Object var, JinjavaInterpreter interpreter, Object[] args, Map<String, Object> kwargs) {
+
+    LinkedHashSet<Object> result = new LinkedHashSet<>();
```
Comment (boulter) [inline src/main/java/com/hubspot/jinjava/lib/filter/UnionFilter.java]: How about using `Sets.union`?

File: src/main/java/com/hubspot/jinjava/lib/filter/SymmetricDifferenceFilter.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    snippets = {
+        @JinjavaSnippet(
+            code = "{{ [1, 2, 3]|symmetric_difference([2, 3, 4]) }}")
+    })
+public class SymmetricDifferenceFilter implements AdvancedFilter {
+
+  @Override
+  public Object filter(Object var, JinjavaInterpreter interpreter, Object[] args, Map<String, Object> kwargs) {
+
+    Set<Object> keysFirst = new HashSet<>();
```
Comment (boulter) [inline src/main/java/com/hubspot/jinjava/lib/filter/SymmetricDifferenceFilter.java]: `Sets.symmetricDifference`?

File: src/main/java/com/hubspot/jinjava/lib/filter/IntersectFilter.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    snippets = {
+        @JinjavaSnippet(
+            code = "{{ [1, 2, 3]|intersect([2, 3, 4]) }}")
+    })
+public class IntersectFilter implements AdvancedFilter {
+
+  @Override
+  public Object filter(Object var, JinjavaInterpreter interpreter, Object[] args, Map<String, Object> kwargs) {
+
+    Set<Object> keys = new HashSet<>();
```
Comment (boulter) [inline src/main/java/com/hubspot/jinjava/lib/filter/IntersectFilter.java]: `Sets.intersection`?

File: src/main/java/com/hubspot/jinjava/lib/filter/DifferenceFilter.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    })
+public class DifferenceFilter implements AdvancedFilter {
+
+  @Override
+  public Object filter(Object var, JinjavaInterpreter interpreter, Object[] args, Map<String, Object> kwargs) {
+
+    Set<Object> keys = new HashSet<>();
+    Set<Object> result = new LinkedHashSet<>();
+
+    Object attr = null;
```
Comment (boulter) [inline src/main/java/com/hubspot/jinjava/lib/filter/DifferenceFilter.java]: `Sets.difference`?

File: src/main/java/com/hubspot/jinjava/lib/filter/UnionFilter.java (hunk 2 above)
Comment (mattcoley) [reply src/main/java/com/hubspot/jinjava/lib/filter/UnionFilter.java]: The reason I did not use the set function helpers is that they would be slower/require more space. I would need to loop over the two iterables to form the sets, then apply those functions. This wouldn't be noticeable for small sets but might be for large sets. The functions are pretty simple anyways, and I'm not using the set helper functions is worth it but it's up to you.

**Comment by mattcoley** [comment]
@boulter Refactored to use the `Sets` functions to simplify the code.

**Pull request merged**